api.js

```diff
@@ -0,0 +1,132 @@
+const {classes: Cc, interfaces: Ci, results: Cr, utils: Cu} = Components;
+
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+XPCOMUtils.defineLazyServiceGetter(this, "loginManager",
+                                   "@mozilla.org/login-manager;1",
+                                   "nsILoginManager");
+XPCOMUtils.defineLazyModuleGetter(this, "NetUtil",
+                                  "resource://gre/modules/NetUtil.jsm");
+
+const FIELDS = {
+  formSubmitURL: "formSubmitURL",
+  origin: "hostname",
+  realm: "httpRealm",
+  username: "username",
+  password: "password",
+  usernameField: "usernameField",
+  passwordField: "passwordField",
+};
+
+function convert(info) {
+  let obj = {};
+  for (let field of Object.keys(FIELDS)) {
+    obj[field] = info[FIELDS[field]];
+  }
+  return obj;
+}
+
+function match(info, search) {
+  return Object.keys(search).every(field => search[field] == null || search[field] == info[FIELDS[field]]);
+}
+
+function accessible(context, origin) {
+  let url;
+  try {
+    url = NetUtil.newURI(origin);
+  } catch (ex) {
+    // unparseable hostname, can this actually happen?
+    return false;
+  }
+
+  if (url.scheme == "addon") {
+    return (url.path == context.extension.id);
+  } else if (url.scheme == "moz-extension") {
+    return (url.host == context.extension.id
+            || url.host == context.extension.uuid);
+  } else {
+    return (context.extension.whiteListedHosts.matches(url));
+  }
+}
+
+class API extends ExtensionAPI {
+  getAPI(context) {
+    // XXX only return this for background contexts?
+    return {
+      logins: {
+        search(query) {
+          let logins = loginManager.getAllLogins()
+              .filter(login => accessible(context, login.hostname))
+              .filter(login => match(login, query))
+              .map(convert);
+
```

Comment:
I have several hundred login entries in my profile.  Since query is a single object, I wonder if using Services.logins.findLogins might be less overhead, filter those results with accessible.